src/openms/source/ANALYSIS/MAPMATCHING/FeatureGroupingAlgorithm.cpp

```diff
@@ -170,6 +191,94 @@ namespace OpenMS
     }
   }
 
+  void FeatureGroupingAlgorithm::annotateIonIdentityNetworks(ConsensusMap& out) const
+  {
+    // bipartite graph with ConsensusFeature indexes and Groups from Features
+    // Vertexes contain uid (index/Group) and is_feature (bool)
+    // e.g. 
+    // ConsensusFeature0 contains Feature (Group = 1) and Feature (Group = 2)
+    // ConsensusFeature1 contains Feature (Group = 2) and Feature (Group = 3)
+    // ConsensusFeature2 contains Feature (Group = 4) and Feature (Group = 5)
+    // graph looks like (Vertex <--> Vertex; component_number)
+    // 0, true <--> 1, false; 0
+    // 0, true <--> 2, false; 0
+    // 1, true <--> 2, false; 0
+    // 1, true <--> 3, false; 0
+    // 2, true <--> 4, false; 1
+    // 2, true <--> 5, false; 1
+    // Total number of components: 2
+    UndirectedOSMIdGraph g;
+
+    for (size_t i = 0; i < out.size(); i++)
+    {
+      out[i].setMetaValue(Constants::UserParam::IIMN_ROW_ID, i+1);
+      if (!out[i].metaValueExists(Constants::UserParam::LINKED_GROUPS)) continue;
+      auto feature_vertex = add_vertex(VertexLabel(String(i), true), g);
+      for (const auto& group: out[i].getMetaValue(Constants::UserParam::LINKED_GROUPS).toStringList())
+      {
+        bool already_set = false;
+        for (auto i : boost::make_iterator_range(vertices(g)))
+        {
+            if ((group == g[i].uid) && (!g[i].is_feature)) 
+            {
+              boost::add_edge(feature_vertex, i, g);
+              already_set = true;
+              break;
+            }
+        }
+        if (!already_set)
+        {
+        boost::add_edge(feature_vertex, add_vertex(VertexLabel(group, false), g), g);
+        }
+      }
+    }
+
+    std::vector<int> components (boost::num_vertices (g));
+    boost::connected_components (g, &components[0]);
+
+    // annotate network number and create a map with feature ID and partner IDs
+    // partner feature vertexes are connected via a group vertex
+    map<size_t, set<size_t>> partner_map;
+    for (auto i : boost::make_iterator_range(vertices(g)))
+    {
+      if (!g[i].is_feature) continue;
+      out[stoi(g[i].uid)].setMetaValue(Constants::UserParam::ANNOTATION_NETWORK_NUMBER, components[i]+1);
+      auto group_neighbours = boost::adjacent_vertices(i, g);
+      for (auto gn : make_iterator_range(group_neighbours))
+      {
+        auto feature_partners = boost::adjacent_vertices(gn, g);
+        for (auto partner : make_iterator_range(feature_partners))
+        {
+          if (g[i].uid == g[partner].uid) continue;
+          partner_map[stoi(g[i].uid)].insert(stoi(g[partner].uid));
+        }
+      }
+    }
+
+    // annotate partners
+    for (const auto& i : partner_map)
+    {
+      for (const auto& j : i.second)
```

Comment:
Probably better to get all the IIMN_ROW_IDs first and set the metavalue only at the end of the inner loop.